### PR TITLE
Set impl fixes

### PR DESCRIPTION
**Pull request by mmatczuk**

### CODE REVIEW

File: internal/set/set.go [hunk 1]
```diff
@@ -148,100 +136,118 @@ func (s *Set) Each(f func(item T) bool) {
...
 		}
+	}
+	if minSize == 0 {
+		return New()
+	}
 
-		for _, set := range sets {
-			if !set.Has(item) {
-				result.Remove(item)
+	t := sets[minPos].Copy()
```
Comment (nussjustin) [inline internal/set/set.go:242]: If len(sets) is 0, minPos will be -1 and this will fail

File: b16set/b16set.go [hunk 2]
```diff
@@ -145,100 +133,118 @@ func (s *Set) Each(f func(item [16]byte) bool) {
 
 // Copy returns a new Set with a copy of s.
 func (s *Set) Copy() *Set {
-	u := New()
+	u := &Set{
```
Comment (dahankzter) [inline b16set/b16set.go]: This makes an assumption on set creation. If we change it we must remember to change everywhere.

File: b16set/b16set.go [hunk 3]
```diff
@@ -145,100 +133,118 @@ func (s *Set) Each(f func(item [16]byte) bool) {
...
-		u.Add(item)
-		return true
-	})
-	for _, set := range sets {
-		set.Each(func(item [16]byte) bool {
-			u.Add(item)
-			return true
-		})
+func Union(sets ...*Set) *Set {
+	maxPos := -1
```
Comment (dahankzter) [inline b16set/b16set.go:190]: Shouldn't maxPos be set to 0 here?

Comment (dahankzter) [reply b16set/b16set.go:190]: I guess it's covered by the next check.

File: internal/set/set.go (hunk 1 above)
Comment (mmatczuk) [reply internal/set/set.go:242]: :+1:

File: b16set/b16set.go (hunk 3 above)
Comment (mmatczuk) [reply b16set/b16set.go:190]: 0 is not good either `if len(sets) == 0`

File: b16set/b16set.go (hunk 2 above)
Comment (mmatczuk) [reply b16set/b16set.go]: Will add a a private constructor.

File: internal/set/set.go (hunk 1 above)
Comment (dahankzter) [reply internal/set/set.go]: There is a check for minSize == 0 above it right @nussjustin ?

Comment (dahankzter) [reply internal/set/set.go:242]: Yes this is a problem. The check for minSize needs to change right?

Comment (mmatczuk) [reply internal/set/set.go:242]: This is handled now by if stmt above.

**Comment by mmatczuk** [comment]
Fixed, PTAL.

File: b16set/b16set.go (hunk 3 above)
Comment (dahankzter) [reply b16set/b16set.go:190]: But the check after catches this so I guess it's fine.

**Pull request merged**